mesa/agent.py

```diff
@@ -57,3 +63,94 @@ def advance(self) -> None:
     @property
     def random(self) -> Random:
         return self.model.random
+
+
+class AgentSet(MutableSet):
+    """Ordered set of agents"""
+
+    def __init__(self, agents: Iterable[Agent], model: Model):
+        self._agents = weakref.WeakKeyDictionary()
+        for agent in agents:
+            self._agents[agent] = None
+
+        self.model = model
+
+    def __len__(self) -> int:
+        return len(self._agents)
+
+    def __iter__(self) -> Iterator[Agent]:
+        return self._agents.keys()
+
+    def __contains__(self, agent: Agent) -> bool:
+        """Check if an agent is in the AgentSet."""
+        return agent in self._agents
+
+    def select(self, filter_func: Callable[[Agent], bool] | None = None, inplace: bool = False) -> AgentSet:
+        if filter_func is None:
+            return self if inplace else AgentSet(list(self._agents.keys()), self.model)
+
+        agents = [agent for agent in self._agents.keys() if filter_func(agent)]
+
+        if inplace:
+            self._reorder(agents)
+            return self
+        else:
+            return AgentSet(
+                agents,
+                self.model
+            )
+
+    def shuffle(self, inplace: bool = False)-> AgentSet:
+        shuffled_agents = list(self._agents.keys())
+        self.model.random.shuffle(shuffled_agents)
+
+        if inplace:
+            self._reorder(shuffled_agents)
+            return self
+        else:
+            return AgentSet(shuffled_agents, self.model)
+
+    def sort(self, key: Callable[[Agent], Any], reverse: bool = False, inplace: bool = False)-> AgentSet:
+        sorted_agents = sorted(self._agents.keys(), key=key, reverse=reverse)
+
+        if inplace:
+            self._reorder(sorted_agents)
+            return self
+        else:
+            return AgentSet(sorted_agents, self.model)
+
+    def _reorder(self, agents: Iterable[Agent]):
+        _agents = weakref.WeakKeyDictionary()
+        for agent in agents:
+            _agents[agent] = None
+
+        self._agents = _agents
+
+    def do_each(self, method_name: str, *args, **kwargs) -> list[Any]:
+        """invoke method on each agent"""
+        return [getattr(agent, method_name)(*args, **kwargs) for agent in self._agents]
+
+    def get_each(self, attr_name: str) -> list[Any]:
+        """get attribute value on each agent"""
+        return [getattr(agent, attr_name) for agent in self._agents]
+
+    def __getitem__(self, item: int | slice) -> Agent:
```

Comment:
Converting all agents to a list seems a bit inefficient for retrieving a single agent

```suggestion
    def __getitem__(self, item: int | slice) -> "Agent":
            if isinstance(item, slice):
                return list(self._agents.keys())[item]
            return next(itertools.islice(self._agents.keys(), item, item+1))
```